sympy/polys/polymatrix.py

```diff
@@ -87,3 +90,126 @@ def _eval_scalar_rmul(self, other):
 
 
 MutablePolyMatrix = PolyMatrix = MutablePolyDenseMatrix
+
+
+class DomainMatrix:
+
+    def __init__(self, rows, shape, domain):
+        self.shape = shape
+        self.rows = [[item for item in row] for row in rows]
```

Comment:
This is conceptually clashing with Matrix.rows. Could it be private (`_rows`, `_lor` for "list of rows", `dat`, etc..? and accessed through public `tolist()`?)

Reply:
Except for this, I squashed all changes aimed at getting the additions I made and the tests to pass into a single commit past your last commit on my domsolve branch. Please feel free to cherry-pick and modify as you see fit.